File: backend/ProductsAPI/Services/ProductService.cs
using ProductsAPI.DTOs;
using ProductsAPI.Models;
using ProductsAPI.Repositories;

namespace ProductsAPI.Services;

public class ProductService : IProductService
{
    private readonly IProductRepository _repository;

    public ProductService(IProductRepository repository)
    {
        _repository = repository;
    }

    public async Task<PagedResult<Product>> GetPagedAsync(int pageNumber, int pageSize, string? search, string? category, decimal? minPrice, decimal? maxPrice)
    {
        return await _repository.GetPagedAsync(pageNumber, pageSize, search, category, minPrice, maxPrice);
    }

    public async Task<Product?> GetByIdAsync(Guid id)
    {
        return await _repository.GetByIdAsync(id);
    }

    public async Task<Product> CreateAsync(CreateProductDto dto)
    {
        var product = new Product
        {
            Id = Guid.NewGuid(),
            Name = dto.Name,
            Description = dto.Description,
            Category = dto.Category,
            Image = dto.Image,
            Price = dto.Price,
            Stock = dto.Stock
        };

        await _repository.AddAsync(product);
        return product;
    }

    public async Task UpdateAsync(UpdateProductDto dto)
    {
        var existingProduct = await _repository.GetByIdAsync(dto.Id);
        if (existingProduct == null)
            return;

        existingProduct.Name = dto.Name;
        existingProduct.Description = dto.Description;
        existingProduct.Category = dto.Category;
        existingProduct.Image = dto.Image;
        existingProduct.Price = dto.Price;
        existingProduct.Stock = dto.Stock;

        await _repository.UpdateAsync(existingProduct);
    }

    public async Task DeleteAsync(Guid id)
    {
        await _repository.DeleteAsync(id);
    }
}

File: frontend/src/app/features/products/services/product.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { Product, PagedResult } from '../../../core/models/product.model';

@Injectable({
    providedIn: 'root'
})
export class ProductService {
    private http = inject(HttpClient);
    private apiUrl = 'http://localhost:5134/api/products';

    getProducts(
        pageNumber: number = 1,
        pageSize: number = 10,
        search?: string,
        category?: string,
        minPrice?: number,
        maxPrice?: number
    ): Observable<PagedResult<Product>> {
        let params = new HttpParams()
            .set('pageNumber', pageNumber)
            .set('pageSize', pageSize);

        if (search) params = params.set('search', search);
        if (category) params = params.set('category', category);
        if (minPrice !== undefined && minPrice !== null) params = params.set('minPrice', minPrice);
        if (maxPrice !== undefined && maxPrice !== null) params = params.set('maxPrice', maxPrice);

        return this.http.get<any>(this.apiUrl, { params }).pipe(
            map((res): PagedResult<Product> => {
                const rawItems = (res.items ?? res.Items ?? []) as any[];
                const items: Product[] = rawItems.map(p => ({
                    id: p.id ?? p.Id,
                    name: p.name ?? p.Name,
                    description: p.description ?? p.Description ?? '',
                    category: p.category ?? p.Category ?? '',
                    image: p.image ?? p.Image ?? '',
                    price: p.price ?? p.Price ?? 0,
                    stock: p.stock ?? p.Stock ?? 0
                }));

                return {
                    items,
                    totalCount: res.totalCount ?? res.TotalCount ?? items.length,
                    pageNumber: res.pageNumber ?? res.PageNumber ?? pageNumber,
                    pageSize: res.pageSize ?? res.PageSize ?? pageSize,
                    totalPages: res.totalPages ?? res.TotalPages ?? 1
                };
            })
        );
    }

    getProductById(id: string): Observable<Product> {
        return this.http.get<any>(`${this.apiUrl}/${id}`).pipe(
            map(p => ({
                id: p.id ?? p.Id,
                name: p.name ?? p.Name,
                description: p.description ?? p.Description ?? '',
                category: p.category ?? p.Category ?? '',
                image: p.image ?? p.Image ?? '',
                price: p.price ?? p.Price ?? 0,
                stock: p.stock ?? p.Stock ?? 0
            }))
        );
    }

    createProduct(product: Omit<Product, 'id'>): Observable<Product> {
        const body = {
            name: product.name,
            description: product.description,
            category: product.category,
            image: product.image,
            price: product.price,
            stock: product.stock
        };

        return this.http.post<any>(this.apiUrl, body).pipe(
            map(p => ({
                id: p.id ?? p.Id,
                name: p.name ?? p.Name,
                description: p.description ?? p.Description ?? '',
                category: p.category ?? p.Category ?? '',
                image: p.image ?? p.Image ?? '',
                price: p.price ?? p.Price ?? 0,
                stock: p.stock ?? p.Stock ?? 0
            }))
        );
    }

    updateProduct(id: string, product: Omit<Product, 'id'>): Observable<void> {
        const body = {
            id,
            name: product.name,
            description: product.description,
            category: product.category,
            image: product.image,
            price: product.price,
            stock: product.stock
        };

        return this.http.put<void>(`${this.apiUrl}/${id}`, body);
    }

    deleteProduct(id: string): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }
}
